### PR TITLE
refactor: port VM charm features, add loki log collection and prometheus alert rules

### PR DESCRIPTION
**Pull request by marcoppenheimer**
Blocked by https://github.com/canonical/zookeeper-k8s-operator/pull/35, which in turn is blocked by https://github.com/canonical/zookeeper-rock/pull/15

##### `refactor: port VM charm features`
- Mountains of tiny formatting diffs
- Bumped lib versions
- Use of `python-libjuju<3.1`
- Pyright static type checks, and fixing various small type warnings
- Type hinting `ZooKeeperK8sCharm`
- Unified `SERVER_JVMFLAGS`
- Avoiding making client connections during cluster tear-down
- Attempt restart on start-hook
- Enabling ssl hostname verification for presenting certs
##### `refactor: tidying up`
- Some stale references to Kafka persisted in docstrings/comments. Now removed
- There were too many `self.peer_relation` methods. Have attempted to focus on going through `self.charm.peer_relation` or `self.charm.app/unit_peer_data` instead
##### `fix: correctly generate SANs on K8s`
- This wasn't addressed prior as we weren't using hostname verification, and there was a bug in the TLS lib
##### `feat: add prometheus alert rules`
- Not sure if the implementation is correct yet. Though https://github.com/canonical/prometheus-k8s-operator/blob/359a9b593ba0f9fae53a660967beedf94952e6a6/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L249-L261 seems to suggest it's OK. Will validate after blocking PR merged

### CODE REVIEW

File: src/prometheus_alert_rules/high_file_descriptors.rule
Comment (marcoppenheimer) [inline src/prometheus_alert_rules/high_file_descriptors.rule:1]: NOTE - All alert rules will need testing

**Review by deusebio — APPROVED** [review]
Changes looks fine. Eventually we should probably migrate COS to use grafana-agent also for the k8s. But let's do this in a dedicated ticket of enhancement / improvements

File: src/prometheus_alert_rules/high_file_descriptors.rule
Comment (marcoppenheimer) [reply src/prometheus_alert_rules/high_file_descriptors.rule:1]: Right now we don't, I'll consult the COS team, though I imagine it's checking that they exist in the UI.

**Review by zmraul — APPROVED** [review]
Neat! Much better now having the relation databag access unified :)

**Pull request merged**